networkx/utils/configs.py

```diff
@@ -0,0 +1,183 @@
+        cls = dataclass(eq=False, slots=True, kw_only=True, match_args=False)(cls)
+        cls._orig_class = orig_class  # Save original class so we can pickle
+        instance = object.__new__(cls)
+        instance.__init__(**kwargs)
+        return instance
+
+    def _check_config(self, key, value):
+        """Check whether config value is valid. This is useful for subclasses."""
+
+    # Control behavior of attributes
+    def __dir__(self):
+        return self.__dataclass_fields__.keys()
+
+    def __setattr__(self, key, value):
+        if key not in self.__dataclass_fields__:
+            raise AttributeError(f"Invalid config name: {key!r}")
+        self._check_config(key, value)
+        object.__setattr__(self, key, value)
+
+    def __delattr__(self, key):
+        raise TypeError(f"Configuration items can't be deleted (can't delete {key!r}).")
+
+    # Be a `collection.abc.Collection`
+    def __contains__(self, key):
+        return key in self.__dataclass_fields__
+
+    def __iter__(self):
+        return iter(self.__dataclass_fields__)
+
+    def __len__(self):
+        return len(self.__dataclass_fields__)
+
+    def __reversed__(self):
+        return reversed(self.__dataclass_fields__)
+
+    # Add dunder methods for `collections.abc.Mapping`
+    def __getitem__(self, key):
+        try:
+            return getattr(self, key)
+        except AttributeError as err:
+            raise KeyError(*err.args) from None
+
+    def __setitem__(self, key, value):
+        try:
+            setattr(self, key, value)
+        except AttributeError as err:
+            raise KeyError(*err.args) from None
+
+    __delitem__ = __delattr__
+    _ipython_key_completions_ = __dir__  # config["<TAB>
+
+    # Go ahead and make it a `collections.abc.Mapping`
+    def get(self, key, default=None):
+        return getattr(self, key, default)
+
+    def items(self):
+        return collections.abc.ItemsView(self)
+
+    def keys(self):
+        return collections.abc.KeysView(self)
+
+    def values(self):
+        return collections.abc.ValuesView(self)
+
+    # dataclass can define __eq__ for us, but do it here so it works after pickling
+    def __eq__(self, other):
+        if not isinstance(other, Config):
+            return NotImplemented
+        return self._orig_class == other._orig_class and self.items() == other.items()
+
+    # Make pickle work
+    def __reduce__(self):
+        return self._deserialize, (self._orig_class, dict(self))
+
+    @staticmethod
+    def _deserialize(cls, kwargs):
+        return cls(**kwargs)
+
+    # Make type annotations work with key and value types; e.g. Config[str, int]
+    def __class_getitem__(cls, item):
+        return types.GenericAlias(cls, item)
+
+
+# Register, b/c `Mapping.__subclasshook__` returns `NotImplemented`
+collections.abc.Mapping.register(Config)
+
+
+class NetworkXConfig(Config):
+    """Write me!"""
+
+    backend_priority: list[str]
+    backends: Config[str, Config]
```

Comment:
`mypy` reminds me again that I still know very little about type annotations. `mypy` doesn't like `Config[str, Config]`.

My "solution" will be to delete `__class_getitem__` so this syntax isn't even possible. The working solution probably includes `Generic` and `TypeVar` and who knows what. Oh well... I thought this was worth trying.